api/envoy/extensions/clusters/dns/v3/dns_cluster.proto

```diff
@@ -0,0 +1,58 @@
+syntax = "proto3";
+
+package envoy.extensions.clusters.dns.v3;
+
+import "google/protobuf/duration.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.clusters.dns.v3";
+option java_outer_classname = "DnsClusterProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/clusters/dns/v3;dnsv3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: DNS cluster configuration]
+
+// Configuration for strict and logical DNS clusters.
+// [#extension: envoy.clusters.strict_dns]
+// [#extension: envoy.clusters.logical_dns]
+
+// [#next-free-field: 7]
+message DnsCluster {
```

Comment:
I think this message also needs a field to indicate whether we want LOGICAL or STRICT semantics.  Otherwise, there's no way for the client to know how the config should behave.

Note that in xDS, the proto message type embedded inside the `google.protobuf.Any` determines which plugin implementation to use -- the name associated with the plugin (e.g., `envoy.cluster.logical_dns`) isn't actually used to select the extension.  So a proto message of type `envoy.extensions.clusters.dns.v3.DnsCluster` can't be used for two different cluster implementations.

Reply:
I'm pretty sure that its the `name` field does indicate which plugin implementation to use. https://github.com/envoyproxy/envoy/blob/c35f0a8f2748b3b582b851933fb42f61688ac729/source/common/upstream/cluster_factory_impl.cc#L46-L56

In line 46, it gets the `name` from the cluster proto and on 56, it gets the corresponding factory by the `name`.

Reply:
Oh, interesting!  I think this is historical behavior, and we should fix it.  Envoy originally used the name to identify which extension to use, but it's switched most of its extension points to instead use the proto message type (see https://github.com/envoyproxy/envoy/pull/9618).  I guess this one just got missed, probably because there really aren't that many cluster type extensions today.

I think we should fix the behavior here as part of this effort.

Reply:
Yes agreed. Either there must be some embedded type, or you need a different names proto wrapper that would indicate the type. The latter IMO is preferable in case the implementation diverges in the future.